src/async.js

```diff
@@ -117,3 +121,37 @@ export const promiseTimeout = (promise, ms) => {
     });
   });
 };
+
+/**
+ * Returns a Promise which will resolve when `eventName` is triggered on `eventEmitter`.
+ * If `checkFn` is specified, it must return truthy for the Promise to resolve.
+ * Example:
+ *   const waitFor123 = waitForEvent(recordProcessor, 'finished', record => record.id === '123');
+ *   # Start processing all the records.
+ *   recordProcessor.startProcessing();
+ *   # Wait for record '123' to be processed.
+ *   const record123 = await waitFor123;
+ * @param {EventEmitter} eventEmitter
+ * @param {string} eventName
+ * @param {function} [checkFn]
+ * @param {number} [timeout]
+ * @returns {Promise<*>}
+ */
+export const waitForEvent = (eventEmitter, eventName, checkFn = noop, timeout = 0) => {
+  assert(eventEmitter);
+  assert(eventEmitter.on);
+  assert(eventName);
+  assert(checkFn);
+
+  const [provider, resolver] = trigger();
```

Comment:
Now that I understand hooks better (I think) I'm wondering to change tigger to:

```
const [value, setValue] = useValue();

setTimeout(() => setValue(100));

await value();
```

Then useValue could take a generic promise and...

```
const [value, setValue] = useValue(onEvent(emitter, 'foo'));

setTimeout(() => emitter.emit('foo'));

await value();
```

Seems more flexible, simpler and composable? (I had been worrying about "trigger").

```
export const onEvent = (eventEmitter, eventName, test) => {
  return new Promise((resolve, reject) => {
    const off = addListener(eventEmitter, eventName, (...args) => {
      const value = test(...args);
      if (value !== undefined) {
        off();
        resolve(value);
      }
    });
  });
}
```

WDYT?

Reply:
Definitely nicer.

Reply:
I don't favor this part:

`const value = test(...args);`

I think a test function should just give a pass/fail result, not potentially manipulate the response.

```
      if (test(...args)) {
        off();
        resolve(...args);
      }
```